lib/npm.js

```diff
@@ -0,0 +1,132 @@
+import _ from 'lodash';
+import path from 'path';
+import { exec } from 'teen_process';
+import { mkdirp, util } from 'appium-support';
+
+const INSTALL_LOCKFILE = '.appium.install.lock';
+const LINK_LOCKFILE = '.appium.link.lock';
+
+export default class NPM {
+
+  constructor (appiumHome) {
+    this.appiumHome = appiumHome;
+  }
+
+  async exec (cmd, args, opts, execOpts = {}) {
+    let { cwd, json, lockFile } = opts;
+    if (!cwd) {
+      cwd = this.appiumHome;
+    }
+    // ensure the directory we want to install inside of exists
+    await mkdirp(cwd);
+    execOpts.cwd = cwd;
+
+    args.unshift(cmd);
+    if (json) {
+      args.push('-json');
+    }
+    let runner = async () => await exec('npm', args, execOpts);
+    if (lockFile) {
+      const acquireLock = util.getLockFileGuard(path.resolve(cwd, lockFile));
```

Comment:
is cwd always writeable? I'd rather use tempdir

Reply:
cwd can't be random, since we have to run the npm commands in specific directories inside of APPIUM_HOME. the user needs to ensure that it's writeable.

Reply:
Does it make sense to have a separate check for the folder writability?

Reply:
Yes, I'll put a check in the config read section, which is always executed at the beginning of any driver subcommand.